man/man1/shmtvaropt0.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SHMTVAROPT0 1"
.TH SHMTVAROPT0 1 "2015-03-10" "SHTOOLS 3.0" "SHTOOLS 3.0"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "SHMTVarOpt0"
.IX Header "SHMTVarOpt0"
.IP "SHMTVarOpt0 \-" 14
.IX Item "SHMTVarOpt0 -"
Calculate the minimum variance and corresponding optimal weights of a localized multitaper spectral estimate using zonal tapers.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 SHMTVarOpt0 (" 25
.IX Item "SUBROUTINE SHMTVarOpt0 ("
L, \s-1TAPERS, LWIN, KMAX, SFF, VAR_OPT, VAR_UNIT, WEIGHT_OPT, UNWEIGHTED_COVAR NOCROSS \s0)
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1TAPERS\s0(\s-1LWIN+1, KMAX\s0), \s-1SFF\s0(L+WIN+1), \s-1VAR_OPT\s0(\s-1KMAX\s0), \s-1VAR_UNIT\s0(\s-1KMAX\s0)
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
L, \s-1LWIN, KMAX\s0
.IP "REAL*8, \s-1OPTIONAL\s0" 19
.IX Item "REAL*8, OPTIONAL"
\&\s-1WEIGHT_OPT\s0(\s-1KMAX, KMAX\s0), \s-1UNWEIGHTED_COVAR\s0(\s-1KMAX,\s0\ \s-1KMAX\s0)
.IP "\s-1INTEGER, OPTIONAL\s0" 19
.IX Item "INTEGER, OPTIONAL"
\&\s-1NOCROSS\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fISHMTVarOpt0\fR will determine the minimum variance that can be achieved by a weighted multitaper spectral analysis using zonal tapers, as is described by Wieczorek and Simons (2006, submitted manuscript). The minimum variance is output as a function of the number of tapers utilized, from 1 to a maximum of \s-1KMAX,\s0 and the corresponding variance using equal weights is output for comparison. The windowing functions are assumed to be solutions to the spherical-cap concentration problem, as determined by a call to \fISHReturnTapersM\fR. The minimum variance and weights are dependent upon the form of the global unwindowed power spectrum, \s-1SFF.\s0
.PP
If the optional argument \s-1WEIGHT_OPT\s0 is specified, then the optimal weights will be returned as a function of the number of tapers employed, from 1 \s-1KMAX.\s0 If \s-1UNWEIGHTED_COVAR\s0 is specified, then the unweighted covariance matrix of the \s-1KMAX\s0 tapers (i.e., F_ij) will be output. If the optional argument \s-1NOCROSS\s0 is set to 1, then only the diagnonal terms of Fij will be computed.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "L" 18
.IX Item "L"
(input) \s-1INTEGER\s0
.Sp
The angular degree to determine the minimum variance and optimal weights.
.IP "\s-1TAPERS\s0" 18
.IX Item "TAPERS"
(input) REAL*8, \s-1DIMENSION \s0(\s-1LWIN+1, KMAX\s0)
.Sp
A matrix of localization functions obtained from \fISHReturnTapersM\fR.
.IP "\s-1LWIN\s0" 18
.IX Item "LWIN"
(input) \s-1INTEGER\s0
.Sp
The spherical harmonic bandwidth of the localizing windows.
.IP "\s-1KMAX\s0" 18
.IX Item "KMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum number of tapers to be used when calculating the minimum variance and optimal weights. All values from 1 to \s-1KMAX\s0 will be returned.
.IP "\s-1SFF\s0" 18
.IX Item "SFF"
(input) REAL*8, \s-1DIMENSION \s0(L+LWIN+1)
.Sp
The global unwindowed power spectrum of the function to be localized.
.IP "\s-1VAR_OPT\s0" 18
.IX Item "VAR_OPT"
(output) REAL*8, \s-1DIMENSION \s0(\s-1KMAX\s0)
.Sp
The minimum variance of the multitaper spectral estimate for degree L using 1 through \s-1KMAX\s0 tapers.
.IP "\s-1VAR_UNIT\s0" 18
.IX Item "VAR_UNIT"
(output) REAL*8, \s-1DIMENSION \s0(\s-1KMAX\s0)
.Sp
The variance of the multitaper spectral estimate using equal weights for degree L using 1 through \s-1KMAX\s0 tapers.
.IP "\s-1WEIGHT_OPT\s0" 18
.IX Item "WEIGHT_OPT"
(output) REAL*8, \s-1OPTIONAL, DIMENSION \s0(\s-1KMAX, KMAX\s0)
.Sp
The optimal weights (in columns) that minimize the multitaper spectral estimate's variance using 1 through \s-1KMAX\s0 tapers.
.IP "\s-1UNWEIGHTED_COVAR\s0" 18
.IX Item "UNWEIGHTED_COVAR"
(output) REAL*8, \s-1OPTIONAL, DIMENSION \s0(\s-1KMAX, KMAX\s0)
.Sp
The unweighted covariance matrix of the \s-1KMAX\s0 tapers (i.e., Fij in Wieczorek and Simons 2007).
.IP "\s-1NOCROSS\s0" 18
.IX Item "NOCROSS"
(input) \s-1INTEGER, OPTIONAL\s0
.Sp
If 1, only the diagonal terms of the covariance matrix Fij will be computed. If 0, all terms will be computed.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIshreturntapers\fR\|(1), \fIshreturntapersm\fR\|(1), \fIshmtvaropt\fR\|(1)
.PP
<http://shtools.ipgp.fr/>
.SH "REFERENCES"
.IX Header "REFERENCES"
Wieczorek, M. A. and F. J. Simons, Minimum variance multitaper spectral estimation on the sphere, \fIJ. Fourier Anal. Appl.\fR, submitted 2006.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2012 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


src/SHMTVarOpt0.f95
Subroutine SHMTVarOpt0(l, tapers, lwin, kmax, Sff, var_opt, var_unit, weight_opt, unweighted_covar, nocross)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	Given the first Kmax tapers of an matrix TAPERS, and an input global power spectrum
!	Sff, this subroutine will compute the variance of the multitaper spectral estimate
!	assuming that the weights are equal to 1/kmax, and by using the optimal weights that
!	minimize the variance. This routine will only work for zonal tapers.
!
!	Calling Parameters
!		IN
!			l		Spherical harmonic degree to compute variances.
!			tapers		An array of (lwin+1,lwin+1) zonal tapers (arranged in columns).
!			lwin		Maximum spherical harmonic degree of the bandlimited tapers.
!			kmax 		Maximum number of tapers to be used in making the
!					spectral estimate.
!			Sff		Known global power spectrum of the unwindowed field.
!		OUT
!			var_opt		Array of dimension kmax containing the variance computed using the
!					optimal weights when using the first 1 to kmax tapers.
!			var_unit	Array of dimension kmax containg the variance computed using
!					equal weights when using the first 1 to kmax tapers.
!		OPTIONAL(out)
!			weight_opt	Matrix of dimension (kmax, kmax) containing the numerical values of the
!					optimal weights to be applied to each spectral estimate (arranged by columns).
!			unweighted_covar	Covariance matrix.
!		OPTIONAL(in)
!			nocross		If present and equal to 1, then the off-diagonal terms of the covariance
!					matrix will be assumed to be zero.
!
!	Dependencies: SHSjkPG0, SSYSV (LAPACK)
!
!	Copyright (c) 2006, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

	use SHTOOLS, only: SHSjkPG0
	
	implicit none
	real*8, intent(in) ::	tapers(:,:), Sff(:)
	real*8, intent(out) ::	var_opt(:), var_unit(:)
	integer, intent(in) ::	l, lwin, kmax
	real*8, intent(out), optional ::	weight_opt(:,:), unweighted_covar(:,:)
	integer, intent(in), optional ::	nocross
	integer, parameter ::	nb = 64
	real*8 ::		Fij(kmax, kmax), ww(kmax), bb(kmax+1), MM(kmax+1, kmax+1), work((kmax+1)*nb)
	integer ::	i,j, m, k, ipiv(kmax+1), info, lwork
	external :: 	DSYSV
	
	if (size(Sff) < l+lwin + 1) then
		print*, "Error --- SHMTVarOpt0"
		print*, "Sff must be dimensioned (L+LWIN+1) where L and LWIN are ", l, lwin
		print*, "Input array is dimensioned ", size(Sff)
		stop
	elseif (size(tapers(:,1)) < lwin+1 .or. size(tapers(1,:)) < kmax) then
		print*, "Error --- SHMTVarOpt0"
		print*, "TAPERS must be dimensioned as (LWIN+1, KMAX) where LWIN and KMAX are ", lwin, kmax
		print*, "Input array is dimensioned ", size(tapers(:,1)), size(tapers(1,:))
		stop
	elseif ( size(var_opt) < kmax) then
		print*, "Error --- SHMTVarOpt0"
		print*, "VAR_OPT must be dimensioned (KMAX) where KMAX is ",  kmax
		print*, "Input array is dimensioned ",  size(var_opt)
		stop
	elseif ( size(var_unit) < kmax) then
		print*, "Error --- SHMTVarOpt0"
		print*, "VAR_UNIT must be dimensioned (KMAX) where KMAX is ",  kmax
		print*, "Input array is dimensioned ",  size(var_unit)
		stop
	endif

	if (present(weight_opt)) then
		if (size(weight_opt(:,1)) < kmax .or. size(weight_opt(1,:)) < kmax) then
			print*, "Error --- SHMTVarOpt0"
			print*, "WEIGHT_OPT must be dimensioned (KMAX, KMAX) where KMAX is ", kmax
			stop
		endif
	endif
	
	if (present(unweighted_covar)) then
		if (size(unweighted_covar(:,1)) < kmax .or. size(unweighted_covar(1,:)) < kmax) then
			print*, "Error --- SHMTVarOpt0"
			print*, "UNWEIGHTED_COVAR must be dimensioned (KMAX, KMAX) where KMAX is ", kmax
			print*, "Input array is dimensioned ", size(unweighted_covar(:,1)), size(unweighted_covar(1,:))
			stop
		endif
	endif
	
	if (present(nocross)) then
		if (nocross .ne. 1 .and. nocross .ne. 0) then
			print*, "Error --- SHMTVarOpt0"
			print*, "NOCROSS must be either 0 (use all elements of covariance matrix) or"
			print*, "1 (set off-diagonal elements of covariance matrix to zero)."
			print*, "Input value is ", nocross
			stop
		endif
	endif
	

	Fij = 0.0d0
	
	lwork = (kmax+1)*nb
	
	var_opt = 0.0d0
	var_unit = 0.0d0
	
	if (present(weight_opt)) then
		weight_opt = 0.0d0
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate matrix Fij, bb
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

	if (present(nocross)) then
	
		if (nocross==1) then
		
			do i=1, kmax, 1
				m=0
				Fij(i,i) = Fij(i,i) + 2.0d0*SHSjkPG0(Sff, i, i, l, m, tapers, lwin)**2
				do m=1, l, 1
					Fij(i,i) = Fij(i,i) + 4.0d0*SHSjkPG0(Sff, i, i, l, m, tapers, lwin)**2
				enddo
			enddo
		
		else
			do i=1, kmax, 1
				do j=i, kmax, 1
					m=0
					Fij(i,j) = Fij(i,j) + 2.0d0*SHSjkPG0(Sff, i, j, l, m, tapers, lwin)**2
					do m=1, l, 1
						Fij(i,j) = Fij(i,j) + 4.0d0*SHSjkPG0(Sff, i, j, l, m, tapers, lwin)**2
					enddo
					if (i/=j) Fij(j,i) = Fij(i,j)
				enddo
			enddo
			
		endif
		
	else

		do i=1, kmax, 1
			do j=i, kmax, 1
				m=0
				Fij(i,j) = Fij(i,j) + 2.0d0*SHSjkPG0(Sff, i, j, l, m, tapers, lwin)**2
				do m=1, l, 1
					Fij(i,j) = Fij(i,j) + 4.0d0*SHSjkPG0(Sff, i, j, l, m, tapers, lwin)**2
				enddo
				if (i/=j) Fij(j,i) = Fij(i,j)
			enddo
		enddo
		
	endif
	
	if (present(unweighted_covar)) then
		unweighted_covar = 0.0d0
		unweighted_covar(1:kmax, 1:kmax) = Fij(1:kmax,1:kmax)
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate variance using equal weights
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	do k=1, kmax
		var_unit(k) = sum(Fij(1:k,1:k)) / dble(k)**2
	enddo

	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate optimal weights and variance using optimal weights
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	do k=1, kmax
		MM(1:k,1:k) = 2.0d0 * Fij(1:k,1:k)
		MM(k+1,1:k) = 1.0d0
		MM(1:k,k+1) = 1.0d0
		MM(k+1,k+1) = 0.0d0
		bb(1:k) = 0.0d0
		bb(k+1) = 1.0d0
		
		call DSYSV("u", k+1, 1, MM, kmax+1, ipiv,  bb,  kmax+1,  work,  lwork, info)
		if (info /=0) then
			print*, "Error --- SHMTVarOpt0"
			print*, "Problem with call to DSYSV."
			print*, "DSYSV info = ", info
			stop
		endif
		
		if (work(1)/(k+1) > nb) then
			print*, "Warning --- SHMTVarOpt0"
			print*, "The optimal size of nb is ", work(1)/(k+1)
			print*, "Present size is ", nb
			print*, "Please consider changing this parameter and recompiling SHTOOLS."
		endif
		
		do i=1, k
			ww(i) = bb(i)
		enddo
		
		if (present(weight_opt)) then
			weight_opt(1:k,k) = ww(1:k)
		endif
		
		do i=1, k
			do j=1, k
				var_opt(k) = var_opt(k) + ww(i)*ww(j)*Fij(i,j)
			enddo
		enddo
	
	enddo
	

end subroutine SHMTVarOpt0
